File: client/src/map.cpp
#include <cassert>
#include <cstdint>
#include <format>

#include "raylib.h"

#include "config.h"
#include "map.h"
#include "state.h"

namespace app {

uint16_t getIndexFromHexCoord(HexCoord hex_coord) {
    return hex_coord.i + hex_coord.j * HEX_COUNT_SQRT;
}

// todo: implement planned strategy (plan.md)
ResourceType initResourceTypeFromHexCoord(uint16_t i, uint16_t j) {
    if (i % (HEX_COUNT_SQRT / 4) == 10 && j % (HEX_COUNT_SQRT / 4) == 4) {
        return ResourceType::Metal;
    } else if (i % (HEX_COUNT_SQRT / 4) == 2 && j % (HEX_COUNT_SQRT / 4) == 12) {
        return ResourceType::Oil;
    }
    return ResourceType::None;
}

void initMap() {
    for (uint16_t i = 0; i < HEX_COUNT_SQRT; i++) {
        for (uint16_t j = 0; j < HEX_COUNT_SQRT; j++) {
            Hex hex{
                .hex_coord = HexCoord{.i = i, .j = j},
                .resource_type = initResourceTypeFromHexCoord(i, j)
            };
            hexes.at(getIndexFromHexCoord(hex.hex_coord)) = hex;
        }
    }
}

Hex &getHexFromHexCoord(std::vector<Hex> &hexes, HexCoord hex_coord) {
    return hexes.at(getIndexFromHexCoord(hex_coord));
}

MapCoord mapCoordFromHexCoord(HexCoord hex_coord) {
    bool even_row = hex_coord.j % 2 == 0;
    float x = (hex_coord.i * HEX_HEIGHT) + (even_row ? 0 : HEX_HEIGHT / 2);
    float y = hex_coord.j * (HEX_RADIUS + HEX_SIDE_LENGTH / 2);
    return MapCoord{
        .x = x,
        .y = y
    };
}

HexCoord hexCoordFromMapCoord(MapCoord map_coord) {
    int16_t j = static_cast<int16_t>((map_coord.y - HEX_SIDE_LENGTH / 2) / (HEX_RADIUS + HEX_SIDE_LENGTH / 2));
    bool const even_row = j % 2 == 0;
    int16_t i = static_cast<int16_t>((map_coord.x - (even_row ? 0 : HEX_HEIGHT / 2)) / HEX_HEIGHT);

    while (i < 0) {
        i += HEX_COUNT_SQRT;
    }
    while (j < 0) {
        j += HEX_COUNT_SQRT;
    }

    return HexCoord{
        .i = static_cast<uint16_t>(i),
        .j = static_cast<uint16_t>(j)
    };
}

RenderCoord renderCoordFromMapCoord(MapCoord map_origin, MapCoord map_coord) {
    float x{map_coord.x - map_origin.x};
    float y{map_coord.y - map_origin.y};

    if (x < -HEX_HEIGHT / 2) {
        x += getMapWidthPixels();
    }
    if (y < -HEX_RADIUS) {
        y += getMapHeightPixels() - HEX_SIDE_LENGTH / 2;
    }

    return RenderCoord{
        .x = x,
        .y = y,
    };
}

void drawMapHex(MapCoord map_origin, HexCoord hex_coord) {
    // { // Colored hex fill for debugging
    //     static bool color_switch{true};
    //     color_switch = !color_switch;
    //     DrawPoly(center_render_coord, HEX_SIDES, HEX_RADIUS, 30.0f, color_switch ? RED : BLUE);
    // }

    MapCoord map_coord = mapCoordFromHexCoord(hex_coord);
    RenderCoord render_coord = renderCoordFromMapCoord(map_origin, map_coord);
    Hex hex = getHexFromHexCoord(hexes, hex_coord);
    Color color = colorFromResourceType(hex.resource_type);

    if (hex.resource_type != ResourceType::None) {
        DrawPoly(render_coord, HEX_SIDES, HEX_RADIUS, HEX_ROTATION, color);
    }
    DrawPolyLinesEx(render_coord, HEX_SIDES, HEX_RADIUS, HEX_ROTATION, 1.0f, HEX_OUTLINE_COLOR);
    DrawText(std::format("({}, {})", hex_coord.i, hex_coord.j).c_str(), render_coord.x, render_coord.y, 10, RAYWHITE);
}

/**
 * {map_origin} Map coordinate pinned to the top left corner of the screen
 */
void drawMap(MapCoord map_origin) {
    int32_t screen_width{GetScreenWidth()}; // is it any faster to call this function only once per frame? or does raylib already include this caching optimization?
    int32_t screen_height{GetScreenHeight()};
    HexCoord origin_hex_coord{hexCoordFromMapCoord(map_origin)};
    HexCoord min_hex_coord{
        .i = static_cast<uint16_t>(origin_hex_coord.i - 1 < 0 ? HEX_COUNT_SQRT - 1 : origin_hex_coord.i - 1),
        .j = static_cast<uint16_t>(origin_hex_coord.j - 1 < 0 ? HEX_COUNT_SQRT - 1 : origin_hex_coord.j - 1)
    };
    HexCoord hex_coord{min_hex_coord};

    uint16_t max_hexes_i = getHexCountWidth(screen_width);
    uint16_t max_hexes_j = getHexCountHeight(screen_height);
    for (uint16_t hexes_drawn_j = 0; hexes_drawn_j <= max_hexes_j + 2; hexes_drawn_j += 1) {
        for (uint16_t hexes_drawn_i = 0; hexes_drawn_i <= max_hexes_i + 2; hexes_drawn_i += 1) {
            drawMapHex(map_origin, hex_coord);

            hex_coord.i += 1;
            if (hex_coord.i >= HEX_COUNT_SQRT) {
                hex_coord.i = 0;
            }
        }

        hex_coord.i = min_hex_coord.i;
        hex_coord.j += 1;
        if (hex_coord.j >= HEX_COUNT_SQRT) {
            hex_coord.j = 0;
        }
    }
}

} // namespace app

File: client/src/map.rs
use crate::config::{
    HEX_COUNT_SQRT, HEX_HEIGHT, HEX_OUTLINE_COLOR, HEX_RADIUS, HEX_ROTATION, HEX_SIDE_LENGTH,
    HEX_SIDES, get_hex_count_height, get_hex_count_width, get_hex_width_pixels,
    get_map_height_pixels, get_map_width_pixels,
};
use crate::state::{
    Facility, FacilityState, FacilityType, Hex, HexCoord, ResourceType, STATE, State,
};
use raylib::color::Color;
use raylib::ffi::{DrawPoly, DrawPolyLinesEx, DrawText, GetScreenHeight, GetScreenWidth, Vector2};
use std::ffi::{CString, c_int};
use std::ops::{Deref, DerefMut};
use std::sync::RwLockWriteGuard;

#[derive(Debug, Copy, Clone)]
pub struct MapCoord(pub Vector2);

impl Deref for MapCoord {
    type Target = Vector2;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for MapCoord {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl Default for MapCoord {
    fn default() -> Self {
        MapCoord::DEFAULT
    }
}

impl MapCoord {
    pub const DEFAULT: MapCoord = MapCoord(Vector2 { x: 0.0, y: 0.0 });

    pub fn hex_coord(&self) -> HexCoord {
        let mut j: i16 = ((self.y - f32::from(HEX_SIDE_LENGTH / 2))
            / f32::from(HEX_RADIUS + HEX_SIDE_LENGTH / 2)) as i16;
        let even_row: bool = j % 2 == 0;
        let mut i: i16 =
            ((self.x - if even_row { 0. } else { *HEX_HEIGHT / 2. }) / *HEX_HEIGHT) as i16;

        while i < 0 {
            i += HEX_COUNT_SQRT as i16;
        }
        while j < 0 {
            j += HEX_COUNT_SQRT as i16;
        }

        HexCoord {
            i: i as u16,
            j: j as u16,
        }
    }

    pub fn render_coord(&self, map_origin: &MapCoord) -> RenderCoord {
        let mut x: f32 = self.x - map_origin.x;
        let mut y: f32 = self.y - map_origin.y;

        if x < -*HEX_HEIGHT / 2. {
            x += get_map_width_pixels();
        }
        if y < -f32::from(HEX_RADIUS) {
            y += get_map_height_pixels();
        }

        RenderCoord(Vector2 { x, y })
    }

    pub fn overflow_adjusted(&mut self) -> Self {
        let map_width_pixels: f32 = get_map_width_pixels();
        let map_height_pixels: f32 = get_map_height_pixels();

        while self.x < 0. {
            self.x += map_width_pixels;
        }
        while self.y < 0. {
            self.y += map_height_pixels;
        }

        while self.x >= map_width_pixels {
            self.x -= map_width_pixels;
        }
        while self.y >= map_height_pixels {
            self.y -= map_height_pixels;
        }

        *self
    }
}

#[derive(Debug, Copy, Clone)]
pub struct RenderCoord(pub Vector2);

impl Deref for RenderCoord {
    type Target = Vector2;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}

impl DerefMut for RenderCoord {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.0
    }
}

impl Default for RenderCoord {
    fn default() -> Self {
        RenderCoord(Vector2 { x: 0.0, y: 0.0 })
    }
}

impl From<RenderCoord> for Vector2 {
    fn from(value: RenderCoord) -> Self {
        value.0
    }
}

pub fn init_map() {
    for i in 0..HEX_COUNT_SQRT {
        for j in 0..HEX_COUNT_SQRT {
            let hex_coord: HexCoord = HexCoord { i, j };
            let resource_type: ResourceType = init_resource_type_from_hex_coord(&hex_coord);
            let hex: Hex = Hex {
                hex_coord,
                resource_type,
            };
            let mut state_g: RwLockWriteGuard<State> =
                STATE.write().expect("global state poisoned");
            let i: usize = hex.hex_coord.map_index();
            state_g.hexes[i] = hex;
        }
    }
}

// todo: implement planned strategy (plan.md)
const fn init_resource_type_from_hex_coord(hex_coord: &HexCoord) -> ResourceType {
    if (hex_coord.i % (HEX_COUNT_SQRT / 4)) == 10 && hex_coord.j % (HEX_COUNT_SQRT / 4) == 4 {
        ResourceType::Metal
    } else if (hex_coord.i % (HEX_COUNT_SQRT / 4) == 2 && hex_coord.j % (HEX_COUNT_SQRT / 4) == 12)
    {
        ResourceType::Oil
    } else {
        ResourceType::None
    }
}

pub fn draw_map(map_origin: &MapCoord) {
    let screen_width: i32 = unsafe { GetScreenWidth() };
    let screen_height: i32 = unsafe { GetScreenHeight() };
    let origin_hex_coord: HexCoord = map_origin.hex_coord();
    let min_hex_coord: HexCoord = HexCoord {
        i: if i32::from(origin_hex_coord.i) - 1 < 0 {
            HEX_COUNT_SQRT - 1
        } else {
            origin_hex_coord.i - 1
        },
        j: if i32::from(origin_hex_coord.j) - 1 < 0 {
            HEX_COUNT_SQRT - 1
        } else {
            origin_hex_coord.j - 1
        },
    };
    let mut hex_coord: HexCoord = min_hex_coord;

    let max_hexes_i: u16 = get_hex_count_width(screen_width as f32);
    let max_hexes_j: u16 = get_hex_count_height(screen_height as f32);
    for _hexes_drawn_j in 0..(max_hexes_j + 2) {
        for _hexes_drawn_i in 0..(max_hexes_i + 2) {
            draw_map_hex(map_origin, &hex_coord);

            hex_coord.i += 1;
            if hex_coord.i >= HEX_COUNT_SQRT {
                hex_coord.i = 0;
            }
        }

        hex_coord.i = min_hex_coord.i;
        hex_coord.j += 1;
        if hex_coord.j >= HEX_COUNT_SQRT {
            hex_coord.j = 0;
        }
    }
}

fn draw_map_hex(map_origin: &MapCoord, hex_coord: &HexCoord) {
    let map_coord: MapCoord = hex_coord.map_coord();
    let render_coord: RenderCoord = map_coord.render_coord(map_origin);
    let Some(hex): Option<Hex> = hex_coord.clone_map_hex() else {
        panic!("Invalid hex coord: {:?}", hex_coord);
    };
    let color: Color = hex.resource_type.color();

    match hex.resource_type {
        ResourceType::None => {}
        _ => unsafe {
            DrawPoly(
                render_coord.into(),
                HEX_SIDES as c_int,
                f32::from(HEX_RADIUS),
                HEX_ROTATION,
                color.into(),
            );
        },
    }
    unsafe {
        DrawPolyLinesEx(
            render_coord.into(),
            HEX_SIDES as c_int,
            f32::from(HEX_RADIUS),
            HEX_ROTATION,
            1.,
            HEX_OUTLINE_COLOR.into(),
        );
    }
}

pub fn draw_players(map_origin: &MapCoord) {
    let state = STATE.read().expect("global state poisoned");
    for player in &state.players {
        for facility in &player.facilities {
            draw_facility(map_origin, facility);
        }
    }
}

fn draw_facility(map_origin: &MapCoord, facility: &Facility) {
    let map_coord: MapCoord = facility.location.map_coord();
    let render_coord: RenderCoord = map_coord.render_coord(map_origin);
    let color: Color = {
        let mut color = Color {
            r: 0xb4,
            g: 0xb4,
            b: 0xb4,
            a: 0xff,
        };
        match facility.facility_state {
            FacilityState::Operating => {}
            FacilityState::Placing => color.a = 0x80,
            FacilityState::Destroyed => color.a = 0xf0,
        }
        color
    };

    match facility.facility_type {
        FacilityType::ControlCenter => unsafe {
            let cstr = CString::new("CC").unwrap();
            DrawText(
                cstr.as_ptr(),
                render_coord.x as i32 - 10,
                render_coord.y as i32 - 10,
                10,
                color.into(),
            );
        },
        FacilityType::MetalExtractor => unsafe {
            let cstr = CString::new("ME").unwrap();
            DrawText(
                cstr.as_ptr(),
                render_coord.x as i32 - 10,
                render_coord.y as i32 - 10,
                10,
                color.into(),
            );
        },
        FacilityType::OilExtractor => unsafe {
            let cstr = CString::new("OE").unwrap();
            DrawText(
                cstr.as_ptr(),
                render_coord.x as i32 - 10,
                render_coord.y as i32 - 10,
                10,
                color.into(),
            );
        },
    }
}
